File: path_planning/cvdetect/obstacle_detect.py
import threading
import cv2
import numpy as np
import math


class DistanceToCamera(object):

    def __init__(self):

        # camera params
        self.alpha = 8.0 * math.pi / 180 # camera angle
        self.u0 = 200
        self.ay = 500

    def calculate(self, v, h, x_shift, image):
        # compute and return the distance from the target point to the camera
        d = h / math.tan(self.alpha + math.atan((v - self.u0) / self.ay))
        if d > 0:
            cv2.putText(image, "%.1fin" % d,
                (image.shape[1] - x_shift, image.shape[0] - 20), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)
        return d


class ObjectDetection(object):

    def __init__(self):
        self.obstacle_detected = False

    def detect(self, cascade_classifier, gray_image, image):

        # x camera coordinate of the target point 'P'
        v = 0   
        
        # detection
        cascade_obj = cascade_classifier.detectMultiScale(
            gray_image,
            scaleFactor=1.1,
            minNeighbors=5,
            minSize=(30, 30),
        )

        # draw a rectangle around the objects
        for (x_pos, y_pos, width, height) in cascade_obj:
            cv2.rectangle(image, (x_pos+5, y_pos+5), (x_pos+width-5, y_pos+height-5), (255, 255, 255), 2)

            # v = y_pos + height - 5
            v = x_pos + width

            cv2.putText(image, 'Obstacle', (x_pos, y_pos-5), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 0, 255), 2)
            
            print('Obstacle detected at ({},{}) of width {}px.'.format(x_pos, y_pos, width))
            
        return v

class VideoStreamHandler(object):

    def __init__(self):
        # Width of Obstacle
        self.binw = 18 # inches

        self.obj_detection = ObjectDetection()

        # cascade classifiers
        self.bin_cascade = cv2.CascadeClassifier('cascade_xml/cascade.xml') #!!!

        self.d_to_camera = DistanceToCamera()
        self.d_bin = 25

        self.start("source/videos/bin_2.wmv")




    # to eventually handle streams but testing on a video
    def start(self, path):

        cap = cv2.VideoCapture(path)

        while cap.isOpened():

            ret, image = cap.read()

            # resize into 320 x 240
            image = cv2.resize(image, (320, 240))

            # grayscale
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

            # object detection
            v_param = self.obj_detection.detect(self.bin_cascade, gray, image)

            # distance measurement
            if v_param > 0:
                d1 = self.d_to_camera.calculate(v_param, self.binw, 300, image)
                self.d_bin = d1

            cv2.imshow('image', image)

            if cv2.waitKey(1) & 0xFF == ord('q'):
                print('Quit')
                break

        cv2.destroyAllWindows()

if __name__ == '__main__':
    VideoStreamHandler()

File: path_planning/cvdetect/obstacle_detect.cpp
#include "opencv2/objdetect.hpp"
#include "opencv2/imgproc.hpp"
#include "opencv2/opencv.hpp"

#include <stdio.h>
#include <math.h>


using namespace std;
using namespace cv;

/** Function Headers */
void detectAndDisplay( Mat frame );

/** Global variables */
String cascade_name;
CascadeClassifier cascade;
String window_name = "Capture - Obstacle Detection";

/** @function main */
int main( int argc, const char** argv )
{
    CommandLineParser parser(argc, argv,
        "{help h||}"
        "{cascade|cascade_xml/cascade.xml|}");

    cascade_name = parser.get<String>("cascade");
    VideoCapture capture;
    Mat frame;

    //-- 1. Load the cascades
    if( !cascade.load( cascade_name ) ){ printf("--(!)Error loading cascade\n"); return -1; };

    //-- 2. Read the video stream
    capture.open( "source/videos/bin_1.wmv" );

    if ( ! capture.isOpened() ) { printf("--(!)Error opening video capture\n"); return -1; }

    while ( capture.read(frame) )
    {
        if( frame.empty() )
        {
            printf(" --(!) No captured frame -- Break!");
            break;
        }

        //-- 3. Apply the classifier to the frame
        detectAndDisplay( frame );

        if( waitKey(25) == 27 ) { break; } // escape
    }
    return 0;
}

int calculateDist (int v, Mat image) {
    double pi = 3.1415926535897;
    double alpha = 8.0 * alpha/180;
    int u0 = 200;
    int ay = 500;

    double wid = 18
    double d;

    d = wid/tan(alpha + atan((v-u0)/ay));
    
    return d;
}

/** @function detectAndDisplay */
void detectAndDisplay( Mat frame )
{
    std::vector<Rect> obst;
    Mat frame_gray;
    Mat res_gray;

    float angle;

    cvtColor( frame, frame_gray, COLOR_BGR2GRAY );
    resize(frame_gray, res_gray, Size(320, 240))
    // equalizeHist( frame_gray, frame_gray );

    //-- Detect obst
    cascade.detectMultiScale( res_gray, obst, 1.1, 5, 0|CASCADE_SCALE_IMAGE, Size(30, 30) );

    for ( size_t i = 0; i < obst.size(); i++ )
    {
        rectangle(frame, obst[i].x, obst[i].y, obst[i].x+obst[i].width, obst[i].y+obst[i].height, Scalar(255,0,255),2  )
        Point center( obst[i].x + obst[i].width/2, obst[i].y + obst[i].height/2 );

        angle = (center.x - 160)/float(320)

        d = calculateDist(obst[i].x + obst[i].width, res_gray)

        if d > 0:
            putText(image, "%.1fin" % d, Point (frame.size().width - 300, frame.size().height[0] - 20), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 255), 2)

        printf("Detected object at (%d, %d). It is %d in away at angle of %d and width %d px.", center.x, center.y, dist, angle, obst[i].width );

    }
    //-- Show what you got
    imshow( window_name, frame );
}